cognite/client/_api/files.py

```diff
@@ -767,12 +769,30 @@ def download(
                 >>> c = CogniteClient()
                 >>> c.files.download(directory="my_directory", id=[1,2,3], external_id=["abc", "def"])
         """
+
         if isinstance(directory, str):
             directory = Path(directory)
+        assert directory.is_dir(), f"{directory} is not a directory"
+
         all_ids = IdentifierSequence.load(id, external_id).as_dicts()
         id_to_metadata = self._get_id_to_metadata_map(all_ids)
-        assert directory.is_dir(), f"{directory} is not a directory"
-        self._download_files_to_directory(directory, all_ids, id_to_metadata)
+
+        if keep_folder_structure:
+            for metadata in id_to_metadata.values():
+                if metadata.directory:
+                    cdf_directory = Path(metadata.directory[1:])  # Making the directory relative
+                else:
+                    cdf_directory = Path("")
+                file_folder = directory / cdf_directory
+
+                file_folder.mkdir(parents=True, exist_ok=True)
+                self._download_files_to_directory(
+                    directory=file_folder,
+                    all_ids=[item for item in all_ids if item.get("id") == metadata.id],
```

Comment:
This breaks if external ids are used.

Also, this doesn't need to be quadratic in time. Build a identifier -> metadata map upfront.